### PR TITLE
Return top-p/top-k sampling mask/nucleas

### PR DESCRIPTION
**Pull request by hari-hm**
Implement keep-sampling-mask / top-p & top-k replay runtime support through request parsing, sampling, scheduling, detokenization, disaggregation metadata, and the model gateway. 

## Accuracy Tests

`test/registered/sampling/test_sampling_mask.py` was added to check the following:
- basic validations: each sampled token has an attached sampling mask, size of each sampling mask <= specified top-k, and the sampled token always exists in the sampling mask.
- verifies under deterministic inference that sampled tokens when `return_sampling_mask=True` are identical to those returned when `return_sampling_mask=False`, the default path.
- validates that the sampling mask and renormalized logprobs are mathematically correct. This is done by querying the server with `return_logprob = True` and `top_logprobs_num = TOP_K` with temperature 1, which returns the unnormalized logprobs of the TOP_K most probable tokens. From here, we can reconstruct the sampling mask (modulo some possible boundary off-by-ones) and compute the renormalized logprobs offline.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #29291142253](https://github.com/sgl-project/sglang/actions/runs/29291142253)
Latest PR Test (Extra): :x: [Run #29291141839](https://github.com/sgl-project/sglang/actions/runs/29291141839)

### CODE REVIEW

**Review by ByronHsu — COMMENTED** [review]
Thanks for adding this. The Python/SRT side looks broadly safe to me, and the feature itself seems useful.

A few requests before this moves forward:

1. Can we rename the PR/title to avoid making this sound DeepSeek V3.2-specific? From the implementation, `return_sampling_mask` is generic sampling metadata. It is not inherently tied to one model.

2. Please add an end-to-end test in the usual server-launch + sample-request style. The smoke script is useful for manual validation, but I’d like a checked-in test that launches a small server/model, sends a `/generate` request with `return_sampling_mask=true`, and verifies.

3. Please revert the `sgl-model-gateway` raw-JSON/extra-body plumbing in this PR. Switching `/generate` from `Json<GenerateRequest>` to `Json<Value>` and threading `route_generate_with_extra` through the router stack feels too hacky for a known first-class field.

   For Miles, let’s make this change directly in [`radixark/sgl-router-for-miles`](https://github.com/radixark/sgl-router-for-miles) instead. That is the router we will use going forward, and it has the protocol structs plus router code in one repo, so we can add `return_sampling_mask` directly to `GenerateRequest` and keep the handler typed.

4. Please clean up the new `scripts/` additions. 

5. After this lands on `main`, please merge/cherry-pick it into `sglang-miles`.

**Comment by hari-hm** [comment]
> Thanks for adding this. The Python/SRT side looks broadly safe to me, and the feature itself seems useful.
> 
> A few requests before this moves forward:
> 
> 1. Can we rename the PR/title to avoid making this sound DeepSeek V3.2-specific? From the implementation, `return_sampling_mask` is generic sampling metadata. It is not inherently tied to one model.
> 2. Please add an end-to-end test in the usual server-launch + sample-request style. The smoke script is useful for manual validation, but I’d like a checked-in test that launches a small server/model, sends a `/generate` request with `return_sampling_mask=true`, and verifies.
> 3. Please revert the `sgl-model-gateway` raw-JSON/extra-body plumbing in this PR. Switching `/generate` from `Json<GenerateRequest>` to `Json<Value>` and threading `route_generate_with_extra` through the router stack feels too hacky for a known first-class field.
>    For Miles, let’s make this change directly in [`radixark/sgl-router-for-miles`](https://github.com/radixark/sgl-router-for-miles) instead. That is the router we will use going forward, and it has the protocol structs plus router code in one repo, so we can add `return_sampling_mask` directly to `GenerateRequest` and keep the handler typed.
> 4. Please clean up the new `scripts/` additions.
> 5. After this lands on `main`, please merge/cherry-pick it into `sglang-miles`.

Thank you for your review! Your suggestions make sense to me, this PR was more so intended to be a proof-of-concept prior to further cleanup. I will work on doing the suggestions over the coming days.

**Comment by hari-hm** [comment]
Hi @ByronHsu, I believe I have addressed the main points of your feedback and have opened a much cleaner routing PR here https://github.com/radixark/sgl-router-for-miles/pull/7 to use serde. I added an e2e server test script as well in place of the old smoke test script. Please let me know if you have additional feedback that I can address -- thanks!

**Review by ByronHsu — APPROVED** [review]

**Comment by ByronHsu** [comment]
lgtm. please turn the pr to "ready for review" and make the ci pass. current lint is failing too

**Comment by hari-hm** [comment]
@ByronHsu it seems most of the remaining failures in CI are flakes / unrelated to my PR. The relevant test `test/registered/sampling/test_sampling_mask.py` passed. I am not sure how people generally resolve these flaky tests, am I just supposed to rerun CI until they pass?

**Comment by guapisolo** [comment]
will help with this PR merge. calling other code owners

File: python/sglang/srt/disaggregation/decode.py [hunk 1]
```diff
@@ -1580,6 +1583,18 @@ def _commit_transfer_to_req(self, decode_req: DecodeRequest):
                     : decode_req.req.logprob.top_logprobs_num
                 ].tolist()
             )
+        if getattr(decode_req.req, "return_sampling_mask", False):
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/decode.py]: Avoid using getattr in SGLang.

File: python/sglang/srt/disaggregation/utils.py [hunk 2]
```diff
@@ -201,9 +201,22 @@ def __init__(
         hidden_size: int,
         hidden_states_dtype: torch.dtype,
         max_top_logprobs_num: int = 128,
+        max_sampling_mask_tokens: int = int(
+            os.getenv("SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS", "0")
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/utils.py]: Avoid using `os.getenv` for SGLang's environments. See `sglang.srt.environ`

File: python/sglang/srt/disaggregation/utils.py [hunk 3]
```diff
@@ -356,6 +398,42 @@ def set_buf(self, req: Req):
                     dtype=torch.int32,
                     device="cpu",
                 )
+        if getattr(req, "return_sampling_mask", False):
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/utils.py]: Also remove the getattr usage here.

File: python/sglang/srt/disaggregation/utils.py [hunk 4]
```diff
@@ -201,9 +201,22 @@ def __init__(
...
         custom_mem_pool: torch.cuda.MemPool = None,
     ):
         self.custom_mem_pool = custom_mem_pool
+        # The sampling-mask support buffer is large (one int32 per kept vocab id
+        # per slot) and is RDMA-registered + transferred for every request, so
+        # it is opt-in: operators set SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS
+        # to the max support size they want carried across the PD handoff. When
+        # disabled we still allocate a 1-wide placeholder so the aux buffer
+        # layout stays identical on both prefill and decode sides.
+        self.enable_sampling_mask = max_sampling_mask_tokens > 0
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/utils.py:237]: Should remove the buffer register when this feature is disabled.

When P/D sides are mismatched, the alignment should be verified on the P/D communication side, e.g., booststrap.

File: python/sglang/srt/disaggregation/utils.py [hunk 5]
```diff
@@ -241,6 +254,15 @@ def __init__(
             self.output_top_logprobs_idx = torch.zeros(
                 (size, max_top_logprobs_num), dtype=torch.int32, device=device
             )
+            self.output_token_sampling_mask_len = torch.zeros(
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/utils.py]: Register these buffers only when enabling.

File: python/sglang/srt/disaggregation/utils.py [hunk 6]
```diff
@@ -356,6 +398,42 @@ def set_buf(self, req: Req):
                     dtype=torch.int32,
                     device="cpu",
                 )
+        if getattr(req, "return_sampling_mask", False):
+            # Sentinel -1: the decode side records None for this handoff token.
+            # Requests that ask for masks are rejected earlier when the sampling
+            # mask buffer is disabled, so this path only represents a missing
+            # mask from the sampler rather than an intentional transport opt-out.
+            self.output_token_sampling_mask_len[req.metadata_buffer_index][0] = -1
+            sampling_masks = getattr(req, "output_token_sampling_mask", None)
```
Comment (hnyls2002) [inline python/sglang/srt/disaggregation/utils.py]: Also here, avoid using getatter.

File: python/sglang/srt/layers/sampler.py [hunk 7]
```diff
@@ -183,9 +189,20 @@ def forward(
...
                 probs = logits
 
+                if return_sampling_mask:
+                    sampling_mask_data = self._compute_sampling_mask_from_probs(
+                        probs, sampling_info
+                    )
                 batch_next_token_ids = self._sample_from_probs(
                     probs, sampling_info, positions, simple_sampling_case
                 )
+                if sampling_mask_data is not None:
```
Comment (hnyls2002) [inline python/sglang/srt/layers/sampler.py]: Combines these two if branches into one

`if return_sampling_mask` & `if sampling_mask_data is not None`

File: python/sglang/srt/layers/sampler.py [hunk 8]
```diff
@@ -263,6 +280,120 @@ def _sample_from_probs(
                 raise ValueError(f"Invalid sampling backend: {backend}")
         return batch_next_token_ids
 
+    def _compute_sampling_mask_from_probs(
+        self, probs: torch.Tensor, sampling_info: SamplingBatchInfo
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Return sorted token ids, sorted probs, keep mask, and raw probs."""
+        vocab_size = probs.shape[-1]
+        max_top_k = int(sampling_info.top_ks.max().item())
```
Comment (hnyls2002) [inline python/sglang/srt/layers/sampler.py]: Can this sync be removed? Verify if this code is in the sampler's critical path; if it is, prebuild the information on the host side.

File: python/sglang/srt/managers/io_struct.py [hunk 9]
```diff
@@ -163,6 +163,29 @@ class GenerateReqInput:
...
+    # Internally, the sampler first writes these values to
+    # `LogitsProcessorOutput.next_token_sampling_mask_idx` and
+    # `next_token_sampling_logprobs`; Scheduler copies them through
+    # `Req.output_token_sampling_*`; TokenizerManager finally places them in
+    # response `meta_info`.
+    #
+    # `output_token_logprobs` are raw full-vocab logprobs. In contrast,
+    # `output_token_sampling_logprobs` are truncated/renormalized logprobs
+    # over `output_token_sampling_mask` and are the right choice for RL
+    # training that replays the rollout sampling distribution.
```
Comment (hnyls2002) [inline python/sglang/srt/managers/io_struct.py]: Remove/simplify these vibe-coding-like comments.

File: python/sglang/srt/managers/io_struct.py [hunk 10]
```diff
@@ -751,8 +780,12 @@ class TokenizedGenerateReqInput(BaseReq):
     top_logprobs_num: int
     # If return logprobs, the token id to return logprob for
     token_ids_logprob: Optional[List[int]]
+    # Whether to return sparse output-token support from top-k/top-p/min-p
+    # sampling. See GenerateReqInput.return_sampling_mask for the returned
+    # meta_info fields and logprob semantics.
+    return_sampling_mask: bool = False
     # Whether to stream output
-    stream: bool
+    stream: bool = False
```
Comment (hnyls2002) [inline python/sglang/srt/managers/io_struct.py]: Do not change the default behavior? Why not put the `return_sampling_mask` after `stream`?

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 11]
```diff
@@ -1894,6 +1899,28 @@ async def _handle_batch_output(
                     recv_obj,
                     i,
                 )
+            if getattr(state.obj, "return_sampling_mask", False):
```
Comment (hnyls2002) [inline python/sglang/srt/managers/tokenizer_manager.py]: Also, stop using this.

**Review by hnyls2002 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/disaggregation/utils.py (hunk 4 above)
Comment (hnyls2002) [reply python/sglang/srt/disaggregation/utils.py:237]: Can add the P/D control plan's verification in the following PRs.

File: python/sglang/srt/layers/sampler.py
Comment (hnyls2002) [inline python/sglang/srt/layers/sampler.py:135]: The sampling mask behavior is not aligned during this if section. Fail fast if this feature is not supported under the `else` conditions.

**Review by hnyls2002 — CHANGES_REQUESTED** [review]

File: test/registered/sampling/test_sampling_mask.py
Comment (hnyls2002) [inline test/registered/sampling/test_sampling_mask.py:1]: Explain why this test is needed and why this test is not a trivial test in the PR's README without using AI. Otherwise, this test is useless.

**Review by hnyls2002 — CHANGES_REQUESTED** [review]

**Comment by hari-hm** [comment]
Hi @hnyls2002 thanks very much for your review. Making some changes here in accordance with your feedback. Will respond to your inline comments with details on what I changed shortly (feel free to wait to re-review until I respond satisfactorily to each of your comments).

File: python/sglang/srt/disaggregation/decode.py (hunk 1 above)
Comment (hari-hm) [reply python/sglang/srt/disaggregation/decode.py]: removed; here and in all the other places you pointed out.

File: python/sglang/srt/disaggregation/utils.py (hunk 2 above)
Comment (hari-hm) [reply python/sglang/srt/disaggregation/utils.py]: done

File: python/sglang/srt/disaggregation/utils.py (hunk 4 above)
Comment (hari-hm) [reply python/sglang/srt/disaggregation/utils.py:237]: mask buffers are now only allocated and registered when the feature is enabled, otherwise get dropped from get_buf_infos and get_buf. Indeed the buffer count can now differ between P and D if SGLANG_DISAGGREGATION_SAMPLING_MASK_MAX_TOKENS is inconsistently set, but I assume this can be added in the following PRs.

File: python/sglang/srt/layers/sampler.py (hunk 8 above)
Comment (hari-hm) [reply python/sglang/srt/layers/sampler.py]: removed; we now track it on the host side.

File: python/sglang/srt/managers/io_struct.py (hunk 9 above)
Comment (hari-hm) [reply python/sglang/srt/managers/io_struct.py]: removed

File: python/sglang/srt/layers/sampler.py
Comment (hari-hm) [reply python/sglang/srt/layers/sampler.py:135]: The else conditions we should fail fast on are `if self.use_ascend_backend` on line 161 and the `elif use_log_softmax_logprob and enable_deterministic and simple_sampling_case` on line 170. The third `# Standard path: do softmax and sample from probs.` path on line 183 is the functional path.

I added a guard on the ascend case here: https://github.com/hari-hm/sglang/blob/b4fc26222cb9682a1e826a15bd49f71623b44328/python/sglang/srt/managers/scheduler.py#L2210-L2217. I moved it upstream so that requests with the ascend backend cannot even enter a batch if return_sampling_mask = True.

simple_sampling_case is set via 
```
simple_sampling_case = (
      not sampling_info.need_top_p_sampling
      and not sampling_info.need_top_k_sampling
      and not sampling_info.need_min_p_sampling
  )
```
so it will be false by construction when return_sampling_mask = True. So we don't need to worry about this case.

File: test/registered/sampling/test_sampling_mask.py
Comment (hari-hm) [reply test/registered/sampling/test_sampling_mask.py:1]: Added to the README. The top_logprobs_num test is probably the most rigorous.

**Comment by hari-hm** [comment]
Hi @hnyls2002 , I believe I have addressed your first pass comments. Please let me know if you have any follow-up comments and I can make the requisite fixes. Thanks again for your time.

**Comment by hnyls2002** [comment]
### CI States

Latest PR Test (Base): :white_check_mark: [Run #29158553156](https://github.com/sgl-project/sglang/actions/runs/29158553156)
Latest PR Test (Extra): :x: [Run #29158552915](https://github.com/sgl-project/sglang/actions/runs/29158552915)

**Review by hnyls2002 — APPROVED** [review]
